src/Domain/LeanCode.TimeProvider/Time.cs

```diff
@@ -2,10 +2,12 @@ namespace LeanCode.TimeProvider;
 
 public static class Time
 {
-    private static ITimeProvider timeProvider = new UtcTimeProvider();
+    private static readonly AsyncLocal<System.TimeProvider?> Provider = new();
 
-    public static DateTime Now => timeProvider.Now;
-    public static DateTimeOffset NowWithOffset => timeProvider.NowWithOffset;
+    public static System.TimeProvider CurrentProvider => Provider.Value ?? System.TimeProvider.System;
 
-    public static void Use(ITimeProvider newProvider) => timeProvider = newProvider;
+    public static DateTime Now => NowWithOffset.UtcDateTime;
+    public static DateTimeOffset NowWithOffset => CurrentProvider.GetUtcNow();
+
+    public static void Use(System.TimeProvider? newProvider) => Provider.Value = newProvider;
```

Comment:
I'm not convinced this is a good place for async locals - what if you really want to change provider globally? 
On the other hand, setting it on startup _should_ be respected and I don't see a use case to change it later, except tests.

Reply:
I don't think you will be able to set this globally if the async local is at this level. The async machine at the start of the generic host program is really tricky, with multiple independent tasks and breaks in the async chain (last time I checked, so couple of years). :(

Reply:
Well, I was thinking about dedicated `AsyncLocalTimeProvider` adapter from the very beginning, so that `Time` has only one static instance.

The question is, do we need to care about async startup issues, given our use case (system provider by default, overridden only for tests)?

Reply:
Wouldn't exposing and actually using a global override be incredibly thread-unsafe in all but the most trivial use cases? 

Okay, how about two fields, one `static TimeProvider` and one `static AsyncLocal<TimeProvider>` with separate methods to replace them, `UseGlobally` and `UseLocally`, where the former fully resets the latter?

Reply:
@marcin-dardzinski The problem with init is that you would set it the the init, and then it won't be propagated to the request handlers, so we can't ignore that. :)

@Saancred it worked like that for the last couple of years without that much of a problem. The async local is only helpful in tests, so only the fake time provider needs that. 

Let's not introduce the async local at the global level, let's add an adapter that Marcin is talking about.

Reply:
I added a wrapper for that, let me know what you're thinking because I might have overengineered it a bit :see_no_evil: